test/integration/workspace/test_workspace.py

```diff
@@ -319,3 +317,76 @@ def test_workspace_build_editables(self):
                                              "EditableBuild")})
         assert "pkga/0.1: WARN: BUILD PKGA!" in c.out
         assert "pkgb/0.1: WARN: BUILD PKGB!" in c.out
+
+
+class TestWorkspaceBuild:
+    def test_dynamic_products(self):
+        c = TestClient(light=True)
+
+        workspace = textwrap.dedent("""\
+            import os
+            name = "myws"
+
+            workspace_folder = os.path.dirname(os.path.abspath(__file__))
+
+            def products():
+                result = []
+                for f in os.listdir(workspace_folder):
+                    if os.path.isdir(os.path.join(workspace_folder, f)):
+                        if f.startswith("product"):
+                            result.append(f)
+                return result
+                    """)
+
+        c.save({"conanws.py": workspace,
+                "lib1/conanfile.py": GenConanfile("lib1", "0.1"),
+                "product_app1/conanfile.py": GenConanfile("app1", "0.1")})
+        c.run("workspace add lib1")
+        c.run("workspace add product_app1")
```

Comment:
One small detail: doing some tests I found that if you do a `conan workspace remove myproduct` the product is removed from the editable list but not from the products list. I think we want to remote the product from the products in the `yaml` too?